File: services/rust/actix/src/routes/handlers/delete.rs
use crate::routes::{errors::AppError, extractors::AppPath};
use actix_web::{HttpResponse, ResponseError, web::Data};
use app::state::AppState;
use std::sync::Mutex;
use tracing::{error, info, instrument, warn};

#[instrument(skip_all)]
pub async fn delete_handler(state: Data<Mutex<AppState>>, AppPath(id): AppPath) -> HttpResponse {
    if let Ok(mut state) = state.lock() {
        if let Some(task) = state.tasks.remove(&id) {
            drop(state);
            info!(
                %id,
                secret = task.secret.len(),
                operation = task.operation.to_string(),
                method = "DELETE",
                "Removed task"
            );
            return HttpResponse::NoContent().finish();
        } else {
            drop(state);
            warn!(%id, method = "DELETE", "Task not found");
            return AppError::TaskNotFound.error_response();
        }
    }

    error!(%id, method = "DELETE", "Poisoned lock");

    AppError::Internal.error_response()
}

File: services/go/gin/routes/handlers/delete.go
package handlers

import (
	"app"
	"gin/routes"
	"log/slog"
	"net/http"
	"strings"

	"github.com/gin-gonic/gin"
)

func DeleteHandler(state *app.AppState) gin.HandlerFunc {
	return func(c *gin.Context) {
		id, err := ParseID(c)
		if err != nil {
			return
		}

		state.Mu.Lock()
		task, ok := state.Tasks[id]

		if !ok {
			state.Mu.Unlock()
			slog.Warn("Task not found", "method", c.Request.Method, "path", c.Request.URL.Path, "id", id.String())
			routes.AppErrors.TaskNotFound.Error(c)
			return
		}

		delete(state.Tasks, id)

		state.Mu.Unlock()

		slog.Info(
			"Removed task",
			"id", id.String(),
			"secret", len(task.Secret),
			"operation", strings.ToLower(string(task.Operation)),
			"method", c.Request.Method,
		)

		c.Status(http.StatusNoContent)
	}
}
